pkg/providers/cloudstack/apibuilder.go

```diff
@@ -0,0 +1,118 @@
+
+	for _, workerNodeGroupConfig := range clusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations {
+		deployment := machineDeployment(clusterSpec, workerNodeGroupConfig,
+			kubeadmConfigTemplates[workerNodeGroupConfig.Name],
+			machineTemplates[workerNodeGroupConfig.Name],
+		)
+		m[workerNodeGroupConfig.Name] = &deployment
+	}
+	return m
+}
+
+func fillMachineTemplateAnnotations(machineConfig *v1alpha1.CloudStackMachineConfigSpec) map[string]string {
+	annotations := make(map[string]string, 0)
+	if machineConfig.DiskOffering != nil {
+		annotations[fmt.Sprintf("mountpath.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.MountPath
+		annotations[fmt.Sprintf("device.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Device
+		annotations[fmt.Sprintf("filesystem.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Filesystem
+		annotations[fmt.Sprintf("label.diskoffering.%s", constants.CloudstackAnnotationSuffix)] = machineConfig.DiskOffering.Label
+	}
+
+	if machineConfig.Symlinks != nil {
+		links := make([]string, 0)
+		for key := range machineConfig.Symlinks {
+			links = append(links, fmt.Sprintf("%s:%s", key, machineConfig.Symlinks[key]))
+		}
+		// sorting for unit test determinism
+		sort.Strings(links)
+		annotations[fmt.Sprintf("symlinks.%s", constants.CloudstackAnnotationSuffix)] = strings.Join(links, ",")
+	}
+
+	return annotations
+}
+
+func setDiskOffering(machineConfig *v1alpha1.CloudStackMachineConfigSpec, template *cloudstackv1.CloudStackMachineTemplate) {
+	if machineConfig.DiskOffering == nil {
+		return
+	}
+
+	template.Spec.Spec.Spec.DiskOffering = cloudstackv1.CloudStackResourceDiskOffering{
+		CloudStackResourceIdentifier: cloudstackv1.CloudStackResourceIdentifier{
+			ID:   machineConfig.DiskOffering.Id,
+			Name: machineConfig.DiskOffering.Name,
+		},
+		CustomSize: machineConfig.DiskOffering.CustomSize,
+		MountPath:  machineConfig.DiskOffering.MountPath,
+		Device:     machineConfig.DiskOffering.Device,
+		Filesystem: machineConfig.DiskOffering.Filesystem,
+		Label:      machineConfig.DiskOffering.Label,
+	}
+}
+
+func CloudStackMachineTemplate(name string, machineConfig *v1alpha1.CloudStackMachineConfigSpec, now anywhereTypes.NowFunc) *cloudstackv1.CloudStackMachineTemplate {
+	template := &cloudstackv1.CloudStackMachineTemplate{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: cloudstackv1.GroupVersion.String(),
+			Kind:       CloudStackMachineTemplateKind,
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        name,
+			Namespace:   constants.EksaSystemNamespace,
+			Annotations: fillMachineTemplateAnnotations(machineConfig),
+		},
+		Spec: cloudstackv1.CloudStackMachineTemplateSpec{
+			Spec: cloudstackv1.CloudStackMachineTemplateResource{
+				ObjectMeta: metav1.ObjectMeta{
+					CreationTimestamp: metav1.Time{
+						Time: now(),
```

Comment:
I tried not including this field but CAPC appeared to reject it with some validation webhook error saying `CreationTimestamp cannot be null` or something along those lines. I don't believe it's actually used anywhere in CAPC, but it has to be set.

Reply:
If it can't be null, then can we set it to an empty struct and have the system fill it in still? Or is that not possible?

Reply:
@maxdrib do you think we can track where that error comes from? I'm ok moving forward with this if it's a blocker but it sounds weird that webhook will reject this. Looking upstream, it says its set by the system:
https://github.com/kubernetes/apimachinery/blob/master/pkg/apis/meta/v1/types.go#L179-L188

Reply:
Is it getting rejected by a CAPC webhook?

Reply:
Correct @chrisdoherty4, it is being rejected by a CAPC webhook when it's not set. I don't remember the exact logs. I could set this to an empty Time object but that'll be misleading since it'll be epoch time I imagine.

Reply:
```
2022-09-26T14:59:10.137-0400	V6	Executing command	{"cmd": "/usr/local/bin/docker exec -i eksa_1664218435037679000 kubectl apply -f - --namespace eksa-system --kubeconfig eksa-drib-b77a195/generated/eksa-drib-b77a195.kind.kubeconfig"}
2022-09-26T14:59:10.832-0400	V9	docker	{"stderr": "error: error validating \"STDIN\": error validating data: unknown object type \"nil\" in CloudStackMachineTemplate.spec.template.metadata.creationTimestamp; if you choose to ignore these errors, turn validation off with --validate=false\n"}
```

Reply:
that doesn't seem to come from the webhook necessarily, it seems more like the api server
Is this a new machine template or are you updating an existing one?
If it's an update and you set the date during creation, that date would be part of the last-applied annotation, hence kubectl would try to remove the date that was already set.

Reply:
It's a new one - this error appears on cluster creation when I exclude the ObjectMeta entirely. I think previously I had seen some actual webhook error from CAPC but I can't find the log now, it's been a few days.

Reply:
This seems to be a serialization issue. For some reason, no matter what I try, it's being serialized as 
```
    metadata:
      creationTimestamp: null
```
which the api server does not like for some reason. This is also how the creationTimestamp appears in the CSMT's top level ObjectMeta (`creationTimestamp: null`) even when it's not explicitly specified. 

When we were using template-cp, the resulting spec did not have ObjectMeta specified here, and CAPC was happy with that. With that being said, I think we have three options to proceed here

1. Remove this ObjectMeta attribute from the CloudStackMachineTemplate.Spec.Spec in CAPC. This will require an API bump in CAPC and does not feel worth it at this time
2. Modify the serialization logic in the yaml package to remove this field, which feels hacky
3. Set the field to a semi-realistic value, acknowledging that it's unused and proceed with the current implementation.

With those options laid out, I'm inclined to proceed with option 3, and also creating a ticket under the CAPC repo to remove this unused ObjectMeta resource.

However, one thing I still don't understand is why the K8s server accepts the top-level creationTimestamp as null but not this lower-level one. The types are defined the same way in https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/blob/main/api/v1beta2/cloudstackmachinetemplate_types.go

Reply:
`kubectl` always outputs with a null creation timestamp. I've never checked if it was intentional or not but I'd imagine there's something akin to a missing `"omitempty"` somewhere.

```
~/Workspace/github.com/aws/eks-anywhere > kubectl create --dry-run=client -o yaml deployment --image=nginx nginx
apiVersion: apps/v1
kind: Deployment
metadata:
  creationTimestamp: null
  labels:
    app: nginx
  name: nginx
spec:
  replicas: 1
  selector:
    matchLabels:
      app: nginx
  strategy: {}
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: nginx
    spec:
      containers:
      - image: nginx
        name: nginx
        resources: {}
status: {}
```

Reply:
@maxdrib could we add a default for this field in the mutation webhook?

Reply:
I tried adding a defaulter to the CAPC webhook but it didn't even invoke the webhook - I think the request gets rejected before it reaches the webhooks by the K8s server

Reply:
I think I figured it out. I found the exact same issue described in https://github.com/bitnami-labs/sealed-secrets/issues/971, and saw the fix in https://github.com/bitnami-labs/sealed-secrets/pull/979/files, which was to make the field nullable. This will require a change in CAPC

Reply:
I think I figured it out. It was the same issue as https://github.com/bitnami-labs/sealed-secrets/issues/971. I've created a PR in CAPC to make the field nullable: https://github.com/kubernetes-sigs/cluster-api-provider-cloudstack/pull/176, so i think we can proceed without the objectmeta being set here.

Reply:
Fixed in the latest CAPC bundle